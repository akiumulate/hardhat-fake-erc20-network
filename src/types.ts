export interface IFakeERC20Network {
	tokens: Token[]
	defaultMintAmount: string
}

export interface Token {
	name: string
	symbol: string
	defaultMintAmount?: string
}

export interface InitialUserData {
	userAddress: string
	initialBalance: string
}
