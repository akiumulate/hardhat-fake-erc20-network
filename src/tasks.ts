import ora from "ora";
import {
    checkLocalhostNetwork,
    getInitialUserData,
    getTaskResultsDisplay,
} from "./utils";

import type { TaskResults } from "./types";
import type { HardhatRuntimeEnvironment } from "hardhat/types";

export async function deployTokens(hre: HardhatRuntimeEnvironment) {
    const { config, ethers } = hre;

    //Check to make sure a network is running on localhost
    await checkLocalhostNetwork(config);

    const { fakeERC20Network } = config;

    //Get all the users on the network
    const accounts = await ethers.getSigners();

    //Get the ERC20 Factory ABI
    const ERC20FakeFactory = await ethers.getContractFactory(
        "ERC20FakeFactory"
    );

    const spinner = ora();

    //Iterate over each token to deploy the ERC20FakeFactory contract on the local network and
    //mint the token for each signer
    const taskResults: TaskResults = {};

    for (let token of fakeERC20Network.tokens) {
        spinner.start(`Deploying: ${token.name} (${token.symbol})`);

        //Get an array for each user and their initial token balance
        const initialMintAmount =
            token.defaultMintAmount ?? fakeERC20Network.defaultMintAmount;
        let initialUsers = getInitialUserData(accounts, initialMintAmount);

        //Deploy the token and wait until it is mined on the local network
        try {
            let contract = await ERC20FakeFactory.deploy(
                token.name,
                token.symbol,
                initialUsers
            );

            await contract.deployTransaction.wait();

            spinner.succeed(
                `Token Deployed: ${token.name} - (${contract.address})`
            );
            taskResults[token.symbol] = contract.address;
        } catch (error) {
            taskResults[token.symbol] = "Failed";
            spinner.fail(`Token Deployment Failed: ${token.name}`);
        }
    }

    // Display task results
    spinner.info(`Tokens Deployed`);
    const results = getTaskResultsDisplay(taskResults);
    console.log(results);
}
