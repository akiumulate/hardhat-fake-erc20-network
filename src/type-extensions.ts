import "hardhat/types/config";
import "hardhat/types/runtime";

import type { FakeERC20Network } from './types';

/*
	Extends the HardhatConfig and HardhatUserConfig with the fakeERC20Network types
*/
declare module "hardhat/types/config" {
	export interface HardhatConfig {
		fakeERC20Network: FakeERC20Network
    }

	export interface HardhatUserConfig {
		fakeERC20Network?: Partial<FakeERC20Network>
	}
}
