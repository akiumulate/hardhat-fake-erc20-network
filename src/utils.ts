import "isomorphic-fetch";
import { HardhatConfig } from "hardhat/types";

export const TASK_NAME = 'deploy-fake-erc20';

/*
	 Checks to make sure that a local instance of HardHat node is running
 */
export async function checkLocalhostNetwork(config: HardhatConfig) {
    //Make sure the "localhost" url is set so we can ping it
    if (!config?.networks?.localhost?.url) {
        throw new Error("No localhost URL");
    }

    // The JSON-RPC server run by the `hardhat node` task returns status=200 with an
    // empty body when the server is sent a request with the method set to "OPTIONS"
    // and refuses the connection if the server is not running
    const response = await fetch(config.networks.localhost.url, {
        method: "OPTIONS",
    }).catch((err) => {
        throw new Error(
            "Can't find the HardHat local server. You must start the server using the `npx hardhat node` command."
        );
    });

    if (response.status !== 200) {
        throw new Error(
            "Did not get the expected status from the local server"
        );
    }

    return true;
}
