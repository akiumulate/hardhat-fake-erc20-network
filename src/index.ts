import { extendConfig, task } from "hardhat/config";
import "@nomiclabs/hardhat-ethers";

import { deployTokens } from "./tasks";
import { defaultSettings, TASK_NAME } from "./utils";

import "./type-extensions";
import type { HardhatConfig, HardhatUserConfig } from "hardhat/types";

extendConfig(
    (config: HardhatConfig, userConfig: Readonly<HardhatUserConfig>) => {
        //If no fakeERC20Network settings object, then add the default settings
        if (!config.fakeERC20Network) {
            config.fakeERC20Network = defaultSettings;
        }

        //If no tokens, add default token
        if (
            !config.fakeERC20Network.tokens ||
            config.fakeERC20Network.tokens.length === 0
        ) {
            config.fakeERC20Network.tokens = defaultSettings.tokens;
        }

        //If no defaultMintAmount, then add it
        if (!config.fakeERC20Network.defaultMintAmount) {
            config.fakeERC20Network.defaultMintAmount =
                defaultSettings.defaultMintAmount;
        }
    }
);

task(
    TASK_NAME,
    "Deploys fake ERC20 tokens to your localhost network"
).setAction(async (_, hre) => {
	deployTokens(hre);
});
