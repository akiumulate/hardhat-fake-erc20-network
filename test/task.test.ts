import sinon from "sinon";
import chai, { expect } from "chai";
import chaiAsPromised from "chai-as-promised";
import { jestSnapshotPlugin } from "mocha-chai-jest-snapshot";
import decache from "decache";
import { resetHardhatContext } from "hardhat/plugins-testing";

chai.use(chaiAsPromised);
chai.use(jestSnapshotPlugin());

import { useEnvironment } from "./helpers";

import { deployTokens } from "../src/tasks";
import * as utils from "../src/utils";
import { ethers, run } from "hardhat";

describe("Task", function () {
    useEnvironment("full-config");

    let getInitialUserDataSPY = sinon.spy(utils, "getInitialUserData");
    let getTaskResultsDisplaySPY = sinon.spy(utils, "getTaskResultsDisplay");
    let ethersSPY = sinon.spy(ethers);

    afterEach(function () {
        sinon.reset();
    });

    after(function () {
        sinon.restore();
    });

	it("Task - Throws if local network not detected and exits", async function () {
        await expect(deployTokens(this.hre)).to.be.rejectedWith(Error);

        // Expect the task to end so the following functions are not called
        expect(ethersSPY.getSigners.called).to.equal(false);
        expect(ethersSPY.getContractFactory.called).to.equal(false);
        expect(getInitialUserDataSPY.called).to.equal(false);
    });

    it("Task - Task completes successfully", async function () {
        const tokenNumber = this.hre.config.fakeERC20Network.tokens.length;

        //Setup stubs and spies
        sinon.stub(console, "log");

        const checkLocalhostNetworkSTUB = sinon
            .stub(utils, "checkLocalhostNetwork")
            .returns(Promise.resolve(true));

        // Run task
		await run(utils.TASK_NAME);

        //Tests
        expect(checkLocalhostNetworkSTUB.called).to.equal(true);
        expect(ethersSPY.getContractFactory.firstCall.args[0]).to.equal(
            "ERC20FakeFactory"
        );
        expect(getInitialUserDataSPY.callCount).to.equal(tokenNumber);
        expect(getTaskResultsDisplaySPY.calledOnce).to.equal(true);
        expect(getTaskResultsDisplaySPY.firstCall.args[0]).toMatchSnapshot();

        checkLocalhostNetworkSTUB.restore();
    });
});
