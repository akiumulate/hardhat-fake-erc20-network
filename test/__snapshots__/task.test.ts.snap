// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task Task - Task completes successfully 1`] = `
Object {
  "FAKE": "0x5FbDB2315678afecb367f032d93F642f64180aa3",
}
`;
